### DISCUSSION TITLE
BUILD FAILED (macOS 13.2.1 using ruby-build 20230330)

### DISCUSSION THREAD

**Discussion by sebademasi10** [rbenv/ruby-build#2178] [Build failures] [answered]
Last 10 log lines:
compiling normalize.c
installing default libraries
compiling modify.c
compiling set_len.c
compiling enc_str_buf_cat.c
compiling new.c
linking shared-object -test-/string.bundle
436 warnings generated.
linking shared-object date_core.bundle
make: *** [build-ext] Error 2

Please, let me know if I can some additional information in order to solve this problem. Thanks in advance 😊

**Comment by abdibrokhim** [comment]
same issue help me

**Comment by lommoral** [comment]
Try the following. 
Install with homebrew: 
`$ brew install ruby`
This should get the dependencies it needs.
Next, uninstall with homebrew: 
`$ brew uninstall ruby`
Finally, try installing with rbenv again:
`$ rbenv install 3.2.2`
Hope this helps

**Reply by abdibrokhim** [reply]
bro, my current ruby is 2.6..... how can i switch to 3.2.2 ?

**Reply by lommoral** [reply]
Based on this [https://github.com/rbenv/rbenv#installation](url), after step 2 (load rbenv: `$ rbenv init`), follow the printed instructions to skip step 2 in the future. If you do, step 3: close and restart the terminal. Now you can use [https://github.com/rbenv/rbenv#installing-ruby-versions](url). Basically, after you install the version you want (`$ rbenv install 3.2.2`), set a default Ruby version globally: `$ rbenv global 3.2.2`.
Hope this helps.

**Reply by MuRanJiangXia** [reply]
run  rbenv install 3.2.2
error：
BUILD FAILED (macOS 13.3.1 using ruby-build 20230615)

Last 10 log lines:
compiling ossl_x509ext.c
compiling ossl_x509name.c
compiling ossl_x509req.c
compiling ossl_x509revoked.c
compiling ossl_x509store.c
installing default openssl libraries
linking shared-object openssl.bundle
1 warning generated.
linking shared-object ripper.bundle
make: *** [build-ext] Error 2

**Comment by mislav** [comment, marked as answer]
Hi, sorry that you're encountering a build error. These things can be frustrating.

I'm not sure what causes this, because there is limited info visible in your report, but you can start by:
1. Ensuring that your ruby-build version is up to date;
2. Adding the build dependencies recommended for your platform https://github.com/rbenv/ruby-build/wiki#suggested-build-environment;
3. Looking at the *full build log*, the location of which is printed as part of your BUILD FAILED error message. In the full build log you can see more than just 10 last lines of output, including potentially the real error message. If you can't find it, you are welcome to paste the full build log as a GitHub Gist.

**Reply by byennen** [reply]
@mislav same issue. just started

```
rbenv install 2.7.6
To follow progress, use 'tail -f /var/folders/m4/rbj4m691293926vq4m36ctrr0000gn/T/ruby-build.20230406194643.6224.log' or pass --verbose
Downloading openssl-1.1.1t.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/8dee9b24bdb1dcbf0c3d1e9b02fb8f6bf22165e807f45adeb7c9677536859d3b
Installing openssl-1.1.1t...
Installed openssl-1.1.1t to /Users/defiw0lf/.rbenv/versions/2.7.6

Downloading ruby-2.7.6.tar.bz2...
-> https://cache.ruby-lang.org/pub/ruby/2.7/ruby-2.7.6.tar.bz2
Installing ruby-2.7.6...

WARNING: ruby-2.7.6 is nearing its end of life.
It only receives critical security updates, no bug fixes.

ruby-build: using readline from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 13.2.1 using ruby-build 20230309)

Inspect or clean up the working tree at /var/folders/m4/rbj4m691293926vq4m36ctrr0000gn/T/ruby-build.20230406194643.6224.ZobX6Q
Results logged to /var/folders/m4/rbj4m691293926vq4m36ctrr0000gn/T/ruby-build.20230406194643.6224.log

Last 10 log lines:
installing default nkf libraries
compiling date_strftime.c
installing default syslog libraries
linking shared-object io/console.bundle
compiling date_strptime.c
installing digest libraries
linking shared-object objspace.bundle
linking shared-object nkf.bundle
linking shared-object date_core.bundle
make: *** [build-ext] Error 2
```

**Reply by mislav** [reply]
@byennen Please read again my previous comment, while paying close attention to the third bullet point.

The line `make: *** [build-ext] Error 2` is just a generic, final failure message and isn't the actual cause of why the build failed. The actual error message is somewhere else in the full log.

**Reply by byennen** [reply]
@mislav 

```
readline.c:1904:37: error: use of undeclared identifier 'username_completion_function'; did you mean 'rl_username_completion_function'?
                                    rl_username_completion_function);
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                    rl_username_completion_function
readline.c:79:42: note: expanded from macro 'rl_username_completion_function'
# define rl_username_completion_function username_completion_function
                                         ^
/opt/homebrew/opt/readline/include/readline/readline.h:494:14: note: 'rl_username_completion_function' declared here
extern char *rl_username_completion_function (const char *, int);
             ^
1 error generated.
```

**Reply by mislav** [reply]
@byennen Thanks; this is the actual error message. I have seen it before, but I never found out what causes it. Which version of readline do you have installed with Homebrew? Check with `brew info readline`

If you look at the `Inspect or clean up the working tree at <DIR>` line of output from ruby-build, there should be a file at `<DIR>/ext/readline/mkmf.log`. If you could paste that whole file as a GitHub gist, that would be great. Thanks!

Finally, if your computer has an Apple Silicon processor (i.e. `arm64` architecture), check that you didn't [previously make any hacks](https://github.com/rbenv/ruby-build/discussions/1853#discussioncomment-1871744) to run Homebrew in `86_64` mode. You could also try `brew reinstall readline` to make sure, assuming the `brew` command itself hasn't been tampered with.

Ref. https://github.com/rbenv/ruby-build/issues/1691#issuecomment-983122764

**Reply by cl-ement05** [reply]
Hello

I'm facing the exact same error and even after clean installing brew etc it still doesn't work...
Any ideas ?

**Reply by mislav** [reply]
@cl-ement05 By exact same error, do you mean `error: use of undeclared identifier 'username_completion_function'...`?

If you have reinstalled brew then, I suspect that it's running native on your CPU architecture and that its readline is up to date. Without further information from you, it's near impossible to guess what went wrong. What is your operating system and which Ruby version are you trying to install? You could also consider pasting the full build log as a GitHub Gist. Thanks

**Comment by blift** [comment]
In my case that was "libyaml" missing dependency, just install via brew

```
brew install libyaml
```

**Reply by zerozeroisland** [reply]
Thanks! That did the trick for me on an M2 Mac Mini running 14.2.1

**Comment by 0zguner** [comment]
```
brew uninstall ruby
```
and 
```
xcode-select --install
```
after that i ran
```
rbenv install 3.2.2
```
Fixed issue i was having. i hope it helps.

**Reply by jcburley** [reply]
Thanks, I did try this — but it didn't solve my issue, unfortunately. (Note that `xcode-select --install` didn't do anything other than say `error: command line tools are already installed...`.

I was about to submit a bug report when I noticed, following up on that xcode message, that Software Update wants to install a new OS (13.6.1). Might as well try that; am running out of things to try at this point.

**Reply by jcburley** [reply]
(Install still fails after the OS upgrade.)

**Reply by danielfrance** [reply]
Hey check out my notes here: https://github.com/danielfrance/Installing-Ruby-On-M1-Macs

I spent all day yesterday trying to get things installed and those were my steps to get things working.  There were some tricks to make sure you're using the correct openssl package based on the version of ruby you're hoping to install AND using the correct homebrew setup for M1 macs

**Reply by monfresh** [reply]
@jcburley If you've tried everything and if Daniel's notes don't work for you either, I built a product that is guaranteed to get you up and running in about 15 minutes. In your case, you will most likely need to use the "reset" mode which will back up your dev setup, then clean it all up in about 1 minute. And then you can reinstall everything from scratch in "normal" mode. Check it out: https://www.rubyonmac.dev/

I'm happy to answer any questions about it.

**Reply by jcburley** [reply]
Thanks @danielfrance — I went through your process, but it didn't seem to fix anything in the end.

Since I'm not running an M1 (or Manything) Mac, just a 2019-era MacBook Pro, I'm not surprised...but it was worth trying!

**Reply by jcburley** [reply]
UPDATE: I finally fixed my issue; `brew doctor` had identified some files, specifically (I think being the main problem) an old `/usr/local/lib/libz.*` file, along with symlinks to it.

Uninstalling brew entirely, reinstalling just brew, running `brew doctor`, and cleaning out everything it warned about (except the two files I was sure I wanted and which had nothing to do with Ruby or libraries), allowed `brew install rbenv` and later `rbenv install 3.2.2` to work.

**Comment by jamesnottidge** [comment]
Good day

I've been trying for quite a while to install ruby 2.2.10 on my system a Mac running macOS ventura 13.5.1

After reading @mislav comment, I went through the log file and this is the error I found

```configuring -test-/string
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [.ext/x86_64-darwin22/enc/encdb.bundle] Error 1
```

For more general context, I have provided the logs around there 

```compiling ./enc/windows_1252.c
configuring -test-/load/dot.dot
configuring -test-/marshal/compat
configuring -test-/marshal/usr
configuring -test-/method
configuring -test-/num2int
configuring -test-/path_to_class
configuring -test-/postponed_job
linking encoding encdb.bundle
configuring -test-/printf
configuring -test-/proc
configuring -test-/rational
configuring -test-/recursion
compiling ./enc/trans/single_byte.c
configuring -test-/st/foreach
configuring -test-/st/numhash
configuring -test-/st/update
Undefined symbols for architecture x86_64:
  "_rb_enc_set_base", referenced from:
      _Init_encdb in encdb.o
  "_rb_enc_set_dummy", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_alias", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_declare", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_dummy", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_replicate", referenced from:
      _Init_encdb in encdb.o
  "_rb_encdb_set_unicode", referenced from:
      _Init_encdb in encdb.o
ld: symbol(s) not found for architecture x86_64
configuring -test-/string
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [.ext/x86_64-darwin22/enc/encdb.bundle] Error 1
make: *** [enc] Error 2
make: *** Waiting for unfinished jobs....
compiling ./enc/trans/utf8_mac.c
configuring -test-/struct
configuring -test-/symbol
configuring -test-/tracepoint
linking transcoder transdb.bundle
configuring -test-/typeddata
compiling ./enc/trans/utf_16_32.c
configuring -test-/vm
configuring -test-/wait_for_single_fd
configuring -test-/win32/console
Failed to configure -test-/win32/console. It will not be installed.
configuring -test-/win32/dln
Failed to configure -test-/win32/dln. It will not be installed.
configuring -test-/win32/fd_setsize
Failed to configure -test-/win32/fd_setsize. It will not be installed.
configuring bigdecimal
Undefined symbols for architecture x86_64:
  "_rb_declare_transcoder", referenced from:
      _Init_transdb in transdb.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [.ext/x86_64-darwin22/enc/trans/transdb.bundle] Error 1
make[1]: *** Waiting for unfinished jobs....
configuring continuation
configuring coverage
configuring date
configuring dbm
make: *** [trans] Error 2
header: ndbm.h
library: libc
configuring digest
configuring digest/bubblebabble
configuring digest/md5
configuring digest/rmd160
configuring digest/sha1
configuring digest/sha2
configuring etc
configuring fcntl
configuring fiber
configuring fiddle
configuring gdbm
configuring io/console
configuring io/nonblock
configuring io/wait
configuring json
configuring json/generator
configuring json/parser
configuring mathn/complex
configuring mathn/rational
configuring nkf
configuring objspace
configuring openssl
configuring pathname
configuring psych
configuring pty
configuring racc/cparse
configuring rbconfig/sizeof
configuring readline
configuring ripper
configuring sdbm
configuring socket
configuring stringio
configuring strscan
configuring syslog
configuring thread
configuring tk
...........
check struct members..
check libraries....
Use ActiveTcl libraries (if available).
Search tclConfig.sh and tkConfig.sh....................................
Valid [tclConfig.sh, tkConfig.sh] are found in [["/opt/miniconda3/lib", "/opt/miniconda3/lib"]]
Use [tclConfig.sh, tkConfig.sh] == ["/opt/miniconda3/lib/tclConfig.sh", "/opt/miniconda3/lib/tkConfig.sh"]

Find Tcl/Tk libraries. Make tcltklib.so which is required by Ruby/Tk.
configuring tk/tkutil
configuring win32
Failed to configure win32. It will not be installed.
configuring win32ole
Failed to configure win32ole. It will not be installed.
configuring zlib
```
Configuring zlib is the last line of the log. I would greatly appreciate assistance @mislav 

Things I have tried include reinstalling gmp
Updating home brew
Uninstalling ruby 3.x that was on my system
Updating Xcode developer tools
Upgrading ruby-build
Reinstalling a C compiler
Setting Ruby Configure options to disable tk and win32ole. the command was 
``RUBY_CONFIGURE_OPTS="--disable-tk --disable-win32ole" CFLAGS="-Wno-error=implicit-function-declaration" rbenv install 2.2.10``

**Reply by mislav** [reply]
Hi, Ruby 2.2 is roughly from 10 years ago. It would be a miracle if you managed to successfully compile it for a modern Mac machine running a recent version of macOS. I would suggest editing your app to upgrade Ruby version to a more recent Ruby that is still receiving updates, such as Ruby 3.x.

**Comment by danielfrance** [comment]
In case others are still having issues, I wrote an installation guide here: https://github.com/danielfrance/Installing-Ruby-On-M1-Macs/

My main issue was that I was still using the version of Homebrew that didnt support M1 Macs.  Once I fixed that, I followed the `ruby-build` documentation (https://github.com/rbenv/ruby-build/wiki#suggested-build-environment) and `rbenv` documentation (https://github.com/rbenv/rbenv)  and things fell into place easily

**Comment by JinYang88** [comment]
Try using sudo to execute the commands, it works for me.

**Reply by monfresh** [reply]
This is highly discouraged and absolutely unnecessary when you have a proper Ruby development environment. I've been using Ruby for over 12 years and not a single time have I had to use `sudo` to install either Ruby or any gem. Here's an article I wrote that explains [Why You Should Never Use sudo to Install Ruby Gems](https://www.moncefbelyamani.com/why-you-should-never-use-sudo-to-install-ruby-gems/).

**Comment by khayyamsaleem** [comment]
For me, on my m2 mac, had to install `libffi` via brew.
